commitrmnocb/outcome.go

```diff
@@ -0,0 +1,316 @@
+		outcome = p.checkForReportTransmission(previousOutcome, consensusObservation)
+
+	default:
+		p.lggr.Warnw("Unexpected state in Outcome", "state", nextState)
+		return outcome.Encode()
+	}
+
+	p.lggr.Infow("Commit Plugin Outcome", "outcome", outcome, "oid", p.nodeID)
+	return outcome.Encode()
+}
+
+// ReportRangesOutcome determines the sequence number ranges for each chain to build a report from in the next round
+func (p *Plugin) ReportRangesOutcome(
+	query Query,
+	consensusObservation ConsensusObservation,
+) Outcome {
+	rangesToReport := make([]ChainRange, 0)
+
+	rmnOnRampMaxSeqNumsMap := make(map[cciptypes.ChainSelector]cciptypes.SeqNum)
+	for _, seqNumChain := range query.RmnOnRampMaxSeqNums {
+		rmnOnRampMaxSeqNumsMap[seqNumChain.ChainSel] = seqNumChain.SeqNum
+	}
+
+	observedOnRampMaxSeqNumsMap := consensusObservation.OnRampMaxSeqNums
+	observedOffRampNextSeqNumsMap := consensusObservation.OffRampNextSeqNums
+
+	for chainSel, offRampNextSeqNum := range observedOffRampNextSeqNumsMap {
+		onRampMaxSeqNum, exists := observedOnRampMaxSeqNumsMap[chainSel]
+		if !exists {
+			continue
+		}
+
+		if rmnOnRampMaxSeqNum, exists := rmnOnRampMaxSeqNumsMap[chainSel]; exists {
+			onRampMaxSeqNum = min(onRampMaxSeqNum, rmnOnRampMaxSeqNum)
+		}
+
+		if offRampNextSeqNum <= onRampMaxSeqNum {
+			chainRange := ChainRange{
+				ChainSel:    chainSel,
+				SeqNumRange: [2]cciptypes.SeqNum{offRampNextSeqNum, onRampMaxSeqNum},
+			}
+			rangesToReport = append(rangesToReport, chainRange)
+		}
+	}
+
+	outcome := Outcome{
+		OutcomeType:             ReportIntervalsSelected,
+		RangesSelectedForReport: rangesToReport,
+	}
+
+	return outcome
+}
+
+// Given a set of observed merkle roots, gas prices and token prices, and roots from RMN, construct a report
+// to transmit on-chain
+func (p *Plugin) buildReport(
+	_ Query,
+	consensusObservation ConsensusObservation,
+) Outcome {
+	roots := maps.Values(consensusObservation.MerkleRoots)
+
+	outcomeType := ReportGenerated
+	if len(roots) == 0 {
+		outcomeType = ReportEmpty
+	}
+
+	outcome := Outcome{
+		OutcomeType:   outcomeType,
+		RootsToReport: roots,
+		GasPrices:     consensusObservation.GasPricesArray(),
+		TokenPrices:   consensusObservation.TokenPricesArray(),
+	}
+
+	return outcome
+}
+
+// checkForReportTransmission checks if the OffRamp has an updated set of max seq nums compared to the seq nums that
+// were observed when the most recent report was generated. If an update to these max seq sums is detected, it means
+// that the previous report has been transmitted, and we output ReportTransmitted to dictate that a new report
+// generation phase should begin. If no update is detected, and we've exhausted our check attempts, output
+// ReportTransmissionFailed to signify we stop checking for updates and start a new report generation phase. If no
+// update is detected, and we haven't exhausted our check attempts, output ReportInFlight to signify that we should
+// check again next round.
```

Comment:
Is it accurate to say that this is mandatory to ensure sequence numbers and nonces are ordered correctly?

Reply:
What is mandatory?

Reply:
Waiting for confirmation that the previous report has completed transmission.

Reply:
Not necessarily. The Commit Plugin always starts each round by reading the OnRamp and OffRamp state, so sane Seq Num ranges will be chosen, and in the case where an in-flight report lands after these seq nums have been chosen, I assume the OffRamp will reject a report generated with these obsolete seq num ranges.